### ISSUE TITLE
No kernel memory limit support  "Debian GNU/Linux 8 (jessie)"

### ISSUE THREAD

**Issue by andromeda979** [docker/docker#22245]
Goodmorning everyone,
sorry for the inconvenience,
I have a problem with this error
I added in / etc / default / grub "GRUB_CMDLINE_LINUX_DEFAULT =" cgroup_enable = memory swapaccount = 1 ""

but it continues to give me the same error

```
WARNING: No kernel memory limit support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
```

Tips?

How can I increase the memory limit?
Many thanks to all of you!

**Comment by andromeda979** [comment]
I tried to verify with the script "check-config.sh"
I got this result


> 
> info: reading kernel config from /boot/config-3.16.0-4-amd64 ...
> 
> Generally Necessary:
> - cgroup hierarchy: properly mounted [/sys/fs/cgroup]
> - CONFIG_NAMESPACES: enabled
> - CONFIG_NET_NS: enabled
> - CONFIG_PID_NS: enabled
> - CONFIG_IPC_NS: enabled
> - CONFIG_UTS_NS: enabled
> - CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled
> - CONFIG_CGROUPS: enabled
> - CONFIG_CGROUP_CPUACCT: enabled
> - CONFIG_CGROUP_DEVICE: enabled
> - CONFIG_CGROUP_FREEZER: enabled
> - CONFIG_CGROUP_SCHED: enabled
> - CONFIG_CPUSETS: enabled
> - CONFIG_MEMCG: enabled
> - CONFIG_KEYS: enabled
> - CONFIG_MACVLAN: enabled (as module)
> - CONFIG_VETH: enabled (as module)
> - CONFIG_BRIDGE: enabled (as module)
> - CONFIG_BRIDGE_NETFILTER: enabled
> - CONFIG_NF_NAT_IPV4: enabled (as module)
> - CONFIG_IP_NF_FILTER: enabled (as module)
> - CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
> - CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
> - CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
> - CONFIG_NF_NAT: enabled (as module)
> - CONFIG_NF_NAT_NEEDED: enabled
> - CONFIG_POSIX_MQUEUE: enabled
> 
> Optional Features:
> - CONFIG_USER_NS: enabled
> - CONFIG_SECCOMP: enabled
> - CONFIG_CGROUP_PIDS: missing
> - CONFIG_MEMCG_KMEM: missing
> - CONFIG_MEMCG_SWAP: enabled
> - CONFIG_MEMCG_SWAP_ENABLED: missing
>     (note that cgroup swap accounting is not enabled in your kernel config, you can enable it by setting boot option "swapaccount=1")
> - CONFIG_RESOURCE_COUNTERS: enabled
> - CONFIG_BLK_CGROUP: enabled
> - CONFIG_IOSCHED_CFQ: enabled
> - CONFIG_BLK_DEV_THROTTLING: enabled
> - CONFIG_CGROUP_PERF: enabled
> - CONFIG_CGROUP_HUGETLB: missing
> - CONFIG_NET_CLS_CGROUP: enabled (as module)
> - CONFIG_CGROUP_NET_PRIO: enabled
> - CONFIG_CFS_BANDWIDTH: missing
> - CONFIG_FAIR_GROUP_SCHED: enabled
> - CONFIG_RT_GROUP_SCHED: missing
> - CONFIG_EXT3_FS: missing
> - CONFIG_EXT3_FS_XATTR: missing
> - CONFIG_EXT3_FS_POSIX_ACL: missing
> - CONFIG_EXT3_FS_SECURITY: missing
>     (enable these ext3 configs if you are using ext3 as backing filesystem)
> - CONFIG_EXT4_FS: enabled (as module)
> - CONFIG_EXT4_FS_POSIX_ACL: enabled
> - CONFIG_EXT4_FS_SECURITY: enabled
> - Network Drivers:
>   - "overlay":
>     - CONFIG_VXLAN: enabled (as module)
> - Storage Drivers:
>   - "aufs":
>     - CONFIG_AUFS_FS: enabled (as module)
>   - "btrfs":
>     - CONFIG_BTRFS_FS: enabled (as module)
>   - "devicemapper":
>     - CONFIG_BLK_DEV_DM: enabled (as module)
>     - CONFIG_DM_THIN_PROVISIONING: enabled (as module)
>   - "overlay":
>     - CONFIG_OVERLAY_FS: missing
>   - "zfs":
>     - /dev/zfs: missing
>     - zfs command: missing
>     - zpool command: missing
>

**Comment by trosenkranz** [comment]
Hello everyone.
I can confirm this behavior after todays update from docker-engine:amd64 1.10.3-0~jessie to 1.11.0-0~jessie (Debian Jessie Server) .

**Comment by HackToday** [comment]
is it ?

> CONFIG_MEMCG_SWAP_ENABLED: missing (note that cgroup swap accounting is not enabled in your kernel config, you can enable it by setting boot option "swapaccount=1"

@trosenkranz  do you it worked in 1.10.3 after upgrade to 1.11.0 it not work anymore ?

**Comment by andromeda979** [comment]
I have already added the option in grub
but it does not work



```
GRUB_DEFAULT=0
GRUB_TIMEOUT=5
GRUB_DISTRIBUTOR=`lsb_release -i -s 2> /dev/null || echo Debian`
GRUB_CMDLINE_LINUX_DEFAULT="cgroup_enable=memory swapaccount=1"
```

```
docker version
Client:
Version:      1.11.0
API version:  1.23
Go version:   go1.5.4
Git commit:   4dc5990
Built:        Wed Apr 13 18:17:17 2016
OS/Arch:      linux/amd64
```

**Comment by trosenkranz** [comment]
@HackToday I had enabled "swapaccount=1" in /etc/default/grub before the update to docker 1.11.0 and there was no warning during the "docker info" call. I confirmed, that the option was set after the upgrade.

After checking my logfiles, i registered, that the kernel was updated too (from /var/log/aptitude: linux-image-3.16.0-4-amd64:amd64 3.16.7-ckt20-1+deb8u4 -> 3.16.7-ckt25-1). So I did a downgrade of the kernel and docker and logged the outputs of the "docker info" command and the "check-config.sh" script.

My results:
Old kernel and old docker:
There was **no** warning about a "missing kernel memory limit support"
New kernel and old docker:
There was **no** warning about a "missing kernel memory limit support"
New kernel and new docker:
Now the warning about the "missing kernel memory limit support" appeared.

The outputs of the "check-config.sh" script were the same in all configurations.

**Comment by trosenkranz** [comment]
By the way, I got the message about
> CONFIG_MEMCG_SWAP_ENABLED: missing (note that cgroup swap accounting is not enabled in your kernel config, you can enable it by setting boot option "swapaccount=1")

too. Before and after the upgrade.

**Comment by HackToday** [comment]
did you grub-update after you change above options ?

**Comment by trosenkranz** [comment]
I did.

**Comment by HackToday** [comment]
Could you show your `docker info` ?

**Comment by trosenkranz** [comment]
docker info
> Containers: 3
>    Running: 1
>    Paused: 0
>    Stopped: 2
> Images: 16
> Server Version: 1.11.0
> Storage Driver: aufs
>    Root Dir: /var/lib/docker/aufs
>    Backing Filesystem: extfs
>    Dirs: 37
>    Dirperm1 Supported: true
> Logging Driver: json-file
> Cgroup Driver: cgroupfs
> Plugins:
>    Volume: local
>    Network: null host bridge
> Kernel Version: 3.16.0-4-amd64
> Operating System: Debian GNU/Linux 8 (jessie)
> OSType: linux
> Architecture: x86_64
> CPUs: 4
> Total Memory: 1.957 GiB
> Name: ***
> ID: ****:****:****:****:****:****:****:****:****:****:****:****
> Docker Root Dir: /var/lib/docker
> Debug mode (client): false
> Debug mode (server): false
> Registry: https://index.docker.io/v1/
> WARNING: No kernel memory limit support
> WARNING: No cpu cfs quota support
> WARNING: No cpu cfs period support

Debian is running in an HyperV virtual machine.

**Comment by andromeda979** [comment]
some idea?

**Comment by trosenkranz** [comment]
Currently not.

I installed the 4.5 kernel for testing purposes, but then the docker service does not start. I think this is a different problem.

I am uncertain, whether the kernel memory limit was working with docker 1.10.3. The reason is, that the `check-config.sh` script returned `CONFIG_MEMCG_SWAP_ENABLED: missing` with all docker and kernel versions. One could check, if the `docker info` command was changed/improved.

**Comment by thaJeztah** [comment]
Looks like `docker info` gets this information through https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/pkg/sysinfo/sysinfo_linux.go#L99-L102

**Comment by thaJeztah** [comment]
Oh, actually, you're right; the detection was there in older versions, but there was no warning printed; I added that recently; https://github.com/docker/docker/commit/747a486b4aac2ebbbb28bd713b9a4a929f89353b

**Comment by trosenkranz** [comment]
I checked my logfiles and observed, that even after setting `swapaccount=1` in `/etc/default/grub` the docker daemon warned me about `Your kernel does not support kernel memory limit.`

I only can conclude that the memory limitation never worked and this is not a bug. IMHO it is caused by the OS and its kernel which simply doesn't support the limitation. (Does someone know the name of the kernel configuration option, then i can check whether it was compiled with the necessary config.)

**Comment by trosenkranz** [comment]
Just realized, that the `check-config.sh` script uses the kernel config to determine which features can be used.

The current debian jessie does not support kernel memory limitation, so this issue can be closed.
Thanks for the help.

**Comment by thaJeztah** [comment]
Thanks! I'll close this. I think your conclusions are correct; and this was never working, only there was no warning printed in older versions (so it wasn't visible that it's not supported)

**Comment by HazCod** [comment]
Is there a way to make this work? Or does it work in other OSs?

**Comment by thaJeztah** [comment]
@HazCod it depends on the kernel that the distro ships with, if works on other distros that ship with a newer kernel

**Comment by viossat** [comment]
I solved the issue by upgrading the kernel from 3.16 to 4.9.
